File: FrontEnd/lib/domain/usecases/auth/refresh_token_use_case.dart
/// @file refresh_token_use_case.dart
/// @description Implementación del caso de uso de auto-refresh del token.
/// Encapsula la decisión "si quedan <7 días para expirar, llamar a /auth/refresh".
/// @module Core
/// @layer Domain
library;

import '../../../core/storage/auth_local_data_source.dart';
import '../../../core/utils/jwt_decoder.dart';
import '../../interfaces/usecases/auth/i_refresh_token_use_case.dart';
import '../../repositories/i_auth_repository.dart';

/// [implements] IRefreshTokenUseCase
/// [dependencies] IAuthRepository, AuthLocalDataSource
class RefreshTokenUseCase implements IRefreshTokenUseCase {
  final IAuthRepository      _repository;
  final AuthLocalDataSource  _local;

  const RefreshTokenUseCase({
    required IAuthRepository     repository,
    required AuthLocalDataSource local,
  })  : _repository = repository,
        _local      = local;

  @override
  Future<bool> execute({double refreshThresholdDays = 7.0}) async {
    final token = await _local.getAccessToken();
    if (token == null || token.isEmpty) return false;

    final daysLeft = jwtDaysUntilExpiry(token);
    // Sin `exp` legible o ya expirado: deja que el backend decida — no
    // intentamos refrescar proactivamente (si daysLeft <= 0 el backend
    // ya rechazará validateToken con 401 y la sesión se limpiará).
    if (daysLeft == null) return false;
    if (daysLeft >= refreshThresholdDays) return false;

    await _repository.refreshToken();
    return true;
  }
}

File: BackEnd/src/domain/usecases/auth/RefreshTokenUseCase.ts
/**
 * @file RefreshTokenUseCase.ts
 * @description Caso de uso de renovación de token JWT. El frontend lo invoca
 * cuando al token actual le quedan menos de 7 días para expirar. Devuelve un
 * token nuevo con expiración fresca (JWT_EXPIRES_IN, 30d por defecto) sin
 * requerir credenciales — basta con que el JWT actual sea válido (verificado
 * por AuthMiddleware antes de llegar aquí).
 * @module Auth
 * @layer Domain
 *
 * @implements {IRefreshTokenUseCase}
 * @injectable
 * @dependencies IUserRepository, JwtService
 */

import { injectable, inject } from 'inversify';
import type { IRefreshTokenUseCase } from '../../interfaces/usecases/auth/IRefreshTokenUseCase.js';
import type { IUserRepository } from '../../repositories/IUserRepository.js';
import type { AuthResponseDTO } from '../../dtos/auth/auth-response.dto.js';
import { JwtService } from '../../../presentation/services/JwtService.js';
import { TYPES } from '../../../core/types.js';
import { NotFoundException } from '../../../core/exceptions/NotFoundException.js';

/**
 * Genera un token JWT nuevo para el usuario indicado.
 *
 * @implements {IRefreshTokenUseCase}
 * @injectable
 * @dependencies IUserRepository, JwtService
 */
@injectable()
export class RefreshTokenUseCase implements IRefreshTokenUseCase {
  constructor(
    @inject(TYPES.IUserRepository) private readonly userRepo: IUserRepository,
    @inject(TYPES.JwtService)      private readonly jwtService: JwtService,
  ) {}

  /**
   * Verifica que el usuario sigue activo y emite un token JWT nuevo.
   * El token expirado o inválido se rechaza antes (en AuthMiddleware) — este
   * use case asume que el caller ya pasó esa verificación.
   *
   * @param userId — Id del usuario extraído del JWT actual.
   * @returns AuthResponseDTO con el nuevo token (JWT_EXPIRES_IN fresco) y
   *          los datos públicos actualizados del usuario.
   * @throws {NotFoundException} Si el usuario no existe o está soft-deleted
   *         (`findById` ya filtra `deletedAt:null`).
   */
  async execute(userId: string): Promise<AuthResponseDTO> {
    const user = await this.userRepo.findById(userId);
    if (!user) {
      throw new NotFoundException('User', userId);
    }

    const token = this.jwtService.sign({
      userId: user.id,
      email:  user.email,
      role:   user.role,
    });

    return { token, user: user.sanitizeForPublic() };
  }
}
